### PR TITLE
Added the solution to check for a binary search tree in Java

### PR DESCRIPTION
**Pull request by Meghna180401**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?

Yes.

### Description

This is the implementation to check if a binary tree is a binary search tree in Java.

### Checklist

- [✅ ] I've read the contribution guidelines.
- [✅ ] I've checked the issue list before deciding what to submit.
- [✅ ] I've edited the `README.md` and link to my code.

### Related Issues or Pull Requests
#2229

### CODE REVIEW

File: Java/ds/CheckBST.java [hunk 1]
```diff
@@ -0,0 +1,227 @@
+package neoAlgo;
```
Comment (HarshCasper) [inline Java/ds/CheckBST.java]: Remove the Package

File: Java/ds/CheckBST.java [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+
+/*This is the solution to create a Binary Tree from user inputs taken level wise and to check if that binary tree is a Binary Search tree.This solution works for duplicate input values of nodes as well.
```
Comment (HarshCasper) [inline Java/ds/CheckBST.java]: Reduce the length of the line. Compact sentences will help the readers to better understand the Code

File: Java/ds/CheckBST.java [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+It is not enough to check if the left node of current node is smaller than it and the right node of current node is larger than it respectively.
+This approach would't work for the following example which is actually not a Binary Search Tree --
+     15
+    /   \
+  12     28
+ /  \
+9    24 
+
+Correct Approach:
+1. For each node we have to check, if maximum value in its left sub tree is smaller than the respective node value and the minimum value in its right sub tree is larger than the respective node value.
```
Comment (HarshCasper) [inline Java/ds/CheckBST.java]: Reduce the length of the line. Compact sentences will help the readers to better understand the Code

File: Java/ds/CheckBST.java [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+	It is not a Binary Search Tree
+	 
+	 */
+	
+}
+
+
+
+
+
```
Comment (HarshCasper) [inline Java/ds/CheckBST.java]: Remove extra lines of code

File: Java/ds/CheckBST.java [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+	Enter left  child of 24 or -1 if there is no left child: -1
+	Enter right child of 24 or -1 if there is no right child: -1
+	
+	Sample Output 2:
+	Checking by Approach 1:
+	It is not a Binary Search Tree
+	Checking by Approach 2:
+	It is not a Binary Search Tree
+	 
+	 */
```
Comment (HarshCasper) [inline Java/ds/CheckBST.java]: Remove additional sample I/O. One does the trick :)

File: Java/ds/CheckBST.java [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+	Enter left  child of 190 or -1 if there is no left child: -1
+	Enter right child of 190 or -1 if there is no right child: -1
+	Enter left  child of 213 or -1 if there is no left child: -1
+	Enter right child of 213 or -1 if there is no right child: -1
+	Enter left  child of 12 or -1 if there is no left child: -1
+	Enter right child of 12 or -1 if there is no right child: -1
+	Enter left  child of 24 or -1 if there is no left child: -1
+	Enter right child of 24 or -1 if there is no right child: -1
+	Enter left  child of 27 or -1 if there is no left child: -1
+	Enter right child of 27 or -1 if there is no right child: -1
```
Comment (HarshCasper) [inline Java/ds/CheckBST.java]: Use smaller I/O samples

File: Java/ds/CheckBST.java [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+			System.out.println("It is not a Binary Search Tree");
+		
+		
+		System.out.println("Checking by Approach 2:");
+		check = checkBST2(root);
+		if(check==true)
+			System.out.println("It is a Binary Search Tree");
+		else
+			System.out.println("It is not a Binary Search Tree");
+	}
```
Comment (HarshCasper) [inline Java/ds/CheckBST.java]: Follow consistent indentation

File: Java/ds/CheckBST.java [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+				temp.right = child;
+			}
+		}
+		sc.close();
+		return root;
+	}
+	
+	
+	//Approach 1:
+	//Function to find the maximum element in the Left Sub Tree of any node : Rightmost node with no child
```
Comment (HarshCasper) [inline Java/ds/CheckBST.java]: Remove multiple approaches. Use the best approach itself

File: Java/ds/CheckBST.java [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+                check = false;
+                break;
+            }
+        }
+        return check;
+    }
+	
+	
+	public static void main(String[] args) {
+		// TODO Auto-generated method stub
```
Comment (HarshCasper) [inline Java/ds/CheckBST.java]: What does this comment mean?

**Review by HarshCasper — CHANGES_REQUESTED** [review]

**Comment by Meghna180401** [comment]
I have updated the CheckBST.java file with the changes requested.
@HarshCasper please review it.

File: Java/ds/CheckBST.java [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+	Enter right child of 28 or -1 if there is no right child: -1
+	Enter left  child of 9 or -1 if there is no left child: -1
+	Enter right child of 9 or -1 if there is no right child: -1
+	Enter left  child of 24 or -1 if there is no left child: -1
+	Enter right child of 24 or -1 if there is no right child: -1
+	
+	Sample Output :
+	It is not a Binary Search Tree
+	 
+	 */
```
Comment (HarshCasper) [inline Java/ds/CheckBST.java]: Shift it after the Code ends. Not within the body. Also, add complexities.

File: Java/ds/CheckBST.java [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+        		boolean check = true;
+        		inorder(root);
+        		for(int i=1;i<arr.size();i++){
+            			if(arr.get(i-1)>=arr.get(i)){
+                			check = false;
+                			break;
+            			}
+        		}
+        		return check;
+    	}
```
Comment (HarshCasper) [inline Java/ds/CheckBST.java]: Indent the Code properly

File: Java/ds/CheckBST.java [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+				Node child = new Node(leftChild);
+				tree.add(child);
+				/* Create leftChildNode of a frontNode */
+				temp.left = child;
+			}
+			System.out.print("Enter right child of " + temp.data +" or -1 if there is no right child: ");
+			int rightChild = sc.nextInt();
+			//if rightChild is not null 
+			if (rightChild != -1) {
+				/* Create rightChild */
```
Comment (HarshCasper) [inline Java/ds/CheckBST.java]: Remove redundant comments

File: Java/ds/CheckBST.java [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+				Node child = new Node(rightChild);
+				tree.add(child);
+				/* Create rightChildNode of a frontNode */
```
Comment (HarshCasper) [inline Java/ds/CheckBST.java]: Remove redundant comments

File: Java/ds/CheckBST.java [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+				temp.right = child;
+			}
+		}
+		sc.close();
+		return root;
+	}
+	
+	//Correct Approach:
```
Comment (HarshCasper) [inline Java/ds/CheckBST.java]: What does `Correct Approach` is? Have you discussed an `Incorrect Approach` as well?

File: Java/ds/CheckBST.java [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+			sc.close();
+			return root;
+		}
+		Queue<Node> tree = new LinkedList<>();
+		tree.add(root);
+		while(!tree.isEmpty()) {
+			Node temp = tree.remove();
+			System.out.print("Enter left  child of " + temp.data +" or -1 if there is no left child: ");
+			int leftChild = sc.nextInt();
+			// if leftChild is not null
```
Comment (HarshCasper) [inline Java/ds/CheckBST.java]: At some places you are using `//..........` while at other places you are using `/*............*/`

Can you please be more consistent?

File: Java/ds/CheckBST.java [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+		inorder(root.right);
+	}
+	//Function to check if a Binary Tree is a Binary Search Tree
+	public static boolean checkBST(Node root) {
+        		if(root==null)
+            			return true;
+        		boolean check = true;
+        		inorder(root);
+        		for(int i=1;i<arr.size();i++){
+            			if(arr.get(i-1)>=arr.get(i)){
```
Comment (HarshCasper) [inline Java/ds/CheckBST.java]: Indent the Code. In fact run a formatter on the whole code

**Review by HarshCasper — CHANGES_REQUESTED** [review]

File: Java/ds/CheckBST.java (hunk 14 above)
Comment (Meghna180401) [reply Java/ds/CheckBST.java]: Yes I have discussed the algorithmic explanation for the incorrect approach on lines 7-15. 
However there is obviously no code for it. So, I will make sure to remove this comment of Correct Approach.

File: Java/ds/CheckBST.java (hunk 10 above)
Comment (Meghna180401) [reply Java/ds/CheckBST.java]: I had already mentioned space and time complexities in lines 17-21 after the algorithmic explanation.
Should I mention it again after the test case ?

**Comment by Meghna180401** [comment]
I have updated CheckBST.java with the required changes.
Kindly review it @HarshCasper .

**Review by HarshCasper — APPROVED** [review]

**Pull request merged**